components/raftstore/src/store/fsm/peer.rs

```diff
@@ -1269,6 +1270,224 @@ where
+        // two force leader in same term.
+        self.fsm.peer.raft_group.raft.pre_vote = false;
+        // trigger vote request to all voters, will check the vote result in `check_force_leader`
+        self.fsm.peer.raft_group.campaign().unwrap();
+        assert!(
+            self.fsm.peer.raft_group.raft.state == StateRole::PreCandidate
+                || self.fsm.peer.raft_group.raft.state == StateRole::Candidate
+        );
+        if !self
+            .fsm
+            .peer
+            .raft_group
+            .raft
+            .prs()
+            .votes()
+            .get(&self.fsm.peer.peer_id())
+            .unwrap()
+        {
+            warn!(
+                "pre force leader failed to campaign";
+                "region_id" => self.fsm.region_id(),
+                "peer_id" => self.fsm.peer_id(),
+            );
+            return;
+        }
+
+        self.fsm.peer.force_leader = Some(ForceLeaderState::PreForceLeader {
+            expected_alive_voter,
+        });
+        self.fsm.has_ready = true;
+    }
+
+    fn on_exit_pre_force_leader(&mut self) {
+        info!(
+            "exit pre force leader state";
+            "region_id" => self.fsm.region_id(),
+            "peer_id" => self.fsm.peer_id(),
+        );
+
+        self.fsm.peer.raft_group.raft.pre_vote = true;
+        self.fsm.peer.force_leader = None;
+    }
+
+    fn on_enter_force_leader(&mut self) {
+        info!(
+            "enter force leader state";
+            "region_id" => self.fsm.region_id(),
+            "peer_id" => self.fsm.peer_id(),
+        );
+        assert!(!self.fsm.peer.is_leader());
+
+        // become candidate first to increase term
+        self.fsm.peer.raft_group.raft.become_candidate();
+        self.fsm.peer.raft_group.raft.become_leader();
+        assert!(self.fsm.peer.is_leader());
+        self.fsm.peer.raft_group.raft.set_check_quorum(false);
+
+        // make sure it's not hibernated
+        self.reset_raft_tick(GroupState::Ordered);
+
+        if let Some(ForceLeaderState::PreForceLeader {
+            expected_alive_voter,
+        }) = self.fsm.peer.force_leader.take()
+        {
+            self.fsm.peer.force_leader = Some(ForceLeaderState::ForceLeader {
+                expected_alive_voter,
+            });
+        } else {
+            unreachable!();
+        }
+        self.fsm.has_ready = true;
+    }
+
+    fn on_exit_force_leader(&mut self) {
+        if self.fsm.peer.force_leader.is_none() {
+            return;
+        }
+
+        info!(
+            "exit force leader state";
+            "region_id" => self.fsm.region_id(),
+            "peer_id" => self.fsm.peer_id(),
+        );
+        self.fsm.peer.force_leader = None;
+        // make sure it's not hibernated
+        self.reset_raft_tick(GroupState::Ordered);
+        // leader lease shouldn't be renewed in force leader state.
+        assert!(self.fsm.peer.leader_lease.is_expired());
+        self.fsm
```

Comment:
How about `MsgHup`?

Reply:
why? `campaign` after this line is to trigger `MsgHup`, here we need to step it down to follower